v2.1/doc/npasswd.1
'\"
'\" Copyright 1998, The University of Texas at Austin ("U. T. Austin").
'\" All rights reserved.
'\"
'\" By using this software the USER indicates that he or she has read,
'\" understood and will comply with the following:
'\"
'\" U. T. Austin hereby grants USER permission to use, copy, modify, and
'\" distribute this software and its documentation for any purpose and
'\" without fee, provided that:
'\"
'\" 1. the above copyright notice appears in all copies of the software
'\"    and its documentation, or portions thereof, and 
'\" 2. a full copy of this notice is included with the software and its
'\"    documentation, or portions thereof, and 
'\" 3. neither the software nor its documentation, nor portions thereof,
'\"    is sold for profit. Any commercial sale or license of this software,
'\"    copies of the software, its associated documentation and/or
'\"    modifications of either is strictly prohibited without the prior
'\"    consent of U. T. Austin. 
'\" 
'\" Title to copyright to this software and its associated documentation
'\" shall at all times remain with U. T. Austin. No right is granted to
'\" use in advertising, publicity or otherwise any trademark, service
'\" mark, or the name of U. T. Austin.
'\" 
'\" This software and any associated documentation are provided "as is,"
'\" and U. T. AUSTIN MAKES NO REPRESENTATIONS OR WARRANTIES, EXPRESSED OR
'\" IMPLIED, INCLUDING THOSE OF MERCHANTABILITY OR FITNESS FOR A PARTICULAR
'\" PURPOSE, OR THAT USE OF THE SOFTWARE, MODIFICATIONS, OR ASSOCIATED
'\" DOCUMENTATION WILL NOT INFRINGE ANY PATENTS, COPYRIGHTS, TRADEMARKS OR
'\" OTHER INTELLECTUAL PROPERTY RIGHTS OF A THIRD PARTY. U. T. Austin, The
'\" University of Texas System, its Regents, officers, and employees shall
'\" not be liable under any circumstances for any direct, indirect, special,
'\" incidental, or consequential damages with respect to any claim by USER
'\" or any third party on account of or arising from the use, or inability
'\" to use, this software or its associated documentation, even if U. T.
'\" Austin has been advised of the possibility of those damages.
'\" 
'\" Submit commercialization requests to: Office of the Executive Vice
'\" President and Provost, U. T. Austin, 201 Main Bldg., Austin, Texas,
'\" 78712, ATTN: Technology Licensing Specialist.
'\"
'\" $Id: npasswd.1,v 1.7 1998/12/10 00:00:00 clyde Exp $ (cc.utexas.edu)
'\"
.TH npasswd 1
.SH NAME
npasswd \- change login password, shell or finger information
.SH SYNOPSIS
.B npasswd
[
.I "platform options"
] [
.B \-\^X
.I options
] [ username ]
.br
.B chfn
[
.I "platform options"
] [
.B \-\^X
.I options
] [ username ]
.br
.B chsh
[
.I "platform options"
] [
.B \-\^X
.I options
] [ username ]
.SH DESCRIPTION
.B Npasswd
changes the login password, login shell or finger information for a user.
It is designed to supplement or replace the standard password change
programs 
.BR passwd , 
.B chfn
and
.BR chsh .
.PP
.B Npasswd
addresses some common deficiencies of standard
password change programs.  It supports local passwd
(and shadow) files, NIS maps, and enhanced security shadow databases
for HP-UX and Digital Unix (aka OSF/1).
.PP
New passwords are subjected to a series of 
stringent guessability checks to decrease the chance of users
choosing vulnerable passwords.
.PP
When called as "\fBchfn\fP",
.B npasswd
changes the finger information associated with your user name or the specified
user name.  
.PP
When called as "\fBchsh\fP",
.B npasswd
changes the login shell of your user name or the specified user name.  
.PP
Only the super-user can change the password, finger information or shell
of another user.
.PP
.SH OPTIONS
.PP
Every version of 
.B passwd
supports a different set of command line options.
.B Npasswd
supports a subset of these for a few platforms.
It attempts to recognize the other vendor-specific options and defer to
the vendor-supplied program to handle them.
This list may be incomplete.
.SH "PLATFORM OPTIONS - SunOS 4"
.PP
The following options are implemented:
.RS
\fB \-\^d \-\^e \-\^n \-\^x\fP
.RE
.PP
The following options are deferred to the vendor programs:
.RS
\fB \-\^a \fP
.RE
.PP
The following options are ignored:
.RS
\fB \-\^F \-\^l \-\^y \fP
.RE
.SH "PLATFORM OPTIONS - Solaris 2"
.PP
The following options are implemented:
.RS
\fB \-\^e \-\^g \-\^r\fP
.RE
.PP
The following options are deferred to the vendor programs:
.RS
\fB \-\^a \-\^d \-\^h \-\^l \-\^n \-\^w \-\^s \-\^x \fP
.RE
.PP
The following options are ignored:
.RS
\fB \-\^D\fP
.RE
.SH "PLATFORM OPTIONS - HPUX"
.PP
The following options are implemented:
.RS
\fB \-\^d \-\^e \-\^n \-\^x \fP
.RE
.SH "PLATFORM OPTIONS - Digital UNIX"
.PP
The following options are implemented:
.RS
\fB \-\^f \-\^s \fP
.RE
.SH "PLATFORM OPTIONS - AIX"
.PP
The following options are implemented:
.RS
\fB \-\^f \-\^s \fP
.RE
.SH "OPTIONS FOR ALL PLATFORMS"
.TP 10
.BI \-\^Xc
Process the configuration file, output the settings and exit.
.TP 10
.BI \-\^XC cf
Verify the correctness of the configuration file
.IR cf .
Exit status is 0 if file was ok, 1 if not. 
.TP 10
.BI \-\^XD N
Set internal debugging level to N.
This option may be restricted to root.
.sp
Debug levels are:
.RS
.TP 5
none
None.
.TP 5
verbose
Mild verbosity.
.TP 5
lookup
Trace user lookup.
.TP 5
update
Trace user updating.
.TP 5
config
Trace configuration file processing.
.TP 5
misc1
Reserved.
.TP 5
misc2
Reserved.
.TP 5
pwcheck
Trace password checking.
.TP 5
detail
More detailed tracing.
.TP 5
all
Turn on all debugging.
.RE
.TP 10
.B \-\^XF
Suppress new password checking.
This option is restricted to root and should be used very sparingly.
.TP 10
.B \-\^Xf
Change finger information.
.TP 10
.B \-\^XI
Allow standard input to be a file or pipe.
This option can be used thusly:
.RS
.nf
(echo old-password
 echo new-password
 echo new-password) | npasswd -XI ...
.fi
.sp
This option is restricted to root.
.RE
.TP 10
.B \-\^Xs
Change login shell.
.TP 10
.B \-\^XV
Print version and patch level information.
.SH FILES
@NPASSWD-HOME@/npasswd \- The executable
.br
@NPASSWD-HOME@/passwd.conf \- Configuration file
.br
@NPASSWD-DICT@ \- Default dictionary collection
.br
@NPASSWD-HIST@ \- Default password history database
.SH "SEE ALSO"
@NPASSWD-UTIL@/history_admin \- password history administration utility
.br
@NPASSWD-DOCS@/index.html \- Npasswd Reference Manual (HTML format)
.SH DIAGNOSTICS
Many and hopefully fairly self-explanatory.
.SH BUGS
A substantial section of the password check code
is very specific to English.
.PP
No attempt has been made to internationalize the program.
.PP
Some platforms (e.g. Sun, Digital Unix) have elaborate security
facilities.  Few of these are used by
.BR npasswd ,
mostly because the APIs are not documented.
.PP
For historical reasons, there is support for two obsolete operating
systems - SunOS 4 and Ultrix 4.
This code will removed from any future revisions.
.SH AUTHOR
Clyde Hoover
.br
Academic Computing Services and Instructional Technology Services
.br
The University of Texas at Austin
.br
c.hoover@cc.utexas.edu
.br
\(co 1998, The University of Texas at Austin. All rights reserved. 
'\"
'\" End $RCSfile: npasswd.1,v $


v2.1/NEWS
Latest npasswd news	$Date: 1999/02/01 00:00:00 $

Patch 05	2/1/99	A passel of little bug fixes
	* Added copyright messages to more modules.
	* A bit of code format cleanup.
	* Do not enable nsswitch support unless the "nsswitch.h" include
	  file is present.  This means that nsswitch.conf will ONLY be
	  supported on Solaris 2.
	  IRIX 6.5 and Redhat Linux 4.2 have /etc/nsswitch.conf but totally
	  different API for support.

	* Properly punt the '-r' option on SunOS 5.
	* Fixed minor Makefile things throughout the distribution.
	* Added the option "-XA" to always prompt for the old password.
	* Fixed lexical password check to properly recoginize telephone numbers.
	* Enhanced debug mode in checkpassword program.
	* Use an internal pager to present help files.
	* Reorganize terminal handling code.
	* Fix -XI option to work right when stdin is a pipe
	* Fix Configure password file list input loop
	* Re-activate some Cracklib rules to be more anal about password
	  checking.

Patch 04	10/15/98	Lots of bug fixes, some feature enhacements
	* Substantial rework of history_admin:
		* Add 'merge' function (see manual page)
		* Add '-i' command line option 
		* Allow '-m' with default database
		* Fix some more annoying bugs
	* Modifications to the password history test suite to use
	  and verify the above changes
	* Clarify some issues with password history depth
		* History depth now means "how many *old* passwords matter"
		* Always retain the more recent password, even if too old
	* Allow "@" to mean the default database name in history
	  configuration directives (and the history_admin '-f' option)
	* Fix the "passwd.MaxRepeat" config directive
	* Fix the "ShadowTolerance" config directive
	* Fix the "PasswdTolerance" config directive
	* Fix the "PasswdMatchTries" config directive
	* Add build-time option "SHORT_FILE_WARN" to control behavior
	  when passwd/shadow temp files come up short
	* Support the '-r' option under SunOS 5
	* Allow change of local passwd even if NIS client (unless SunOS 4)
	* More run-time trace messages in various places

Patch 03	9/24/98
	* Fixes the "Cannot talk to NIS server" problem under Solaris 2. 

Patch 02	9/22/98
	* Fixes the build problems on Digital UNIX 4.0d.

Patch 01	9/18/98
	* Fixes build problems when NIS support is not enabled.
	* Fixes a problem with the configuration directive "PrintOnly", which
	  was renamed "PrintableOnly" during the last development cycle but
	  the change did not get propigated everywhere.

Platform status:

AIX 4.1		Builds, installs, changes local passwords
SunOS 4.1.4	Builds, installs, changes local and NIS passwords 
SunOS 5.5	Builds, installs, changes local and NIS passwords 
IRIX 6.1	Builds, installs, changes local passwords 
Digital UNIX 4.0	Builds, installs



npasswd-1.X/npasswd.c

/* --------------------------------------------------------------------  */
/*                                                                       */
/*                         Author: Clyde Hoover                          */
/*                          Computation Center                           */
/*                   The University of Texas at Austin                   */
/*                          Austin, Texas 78712                          */
/*                         clyde@emx.utexas.edu                          */
/*                   uunet!cs.utexas.edu!ut-emx!clyde                    */
/*                                                                       */
/*This code may be distributed freely, provided this notice is retained. */
/*                                                                       */
/* --------------------------------------------------------------------  */
/*
 *	This program duplicates the manual page behavior of the 4.XBSD
 *	passwd(1) command.  It can be configured for use with a variety
 *	of passwd systems (/etc/passwd, /etc/shadow, databases).
 *
 *	The System V support is untested (by the author - other sites
 *	tell me it works).
 *
 *	Here we have only the most abstract data needed (login name,
 *	user id, current password, new password).
 *	All other information needed (the full password line, etc),
 *	is kept down in the 'method' routines.
 *
 *	The 'method' routines are:
 *
 *	pw_initalize()		Do initializations 
 *	pw_getuserbyname()	Get user information by name
 *	pw_permission()		Check if user has permission
 *				 to change this users' password
 *	pw_compare()		Compare passwords
 *	pw_check()		Check password
 *				 Returns 1 if ok, 0 otherwise
 *	pw_replace()		Replace the password
 *	pw_cleanup()		Cleanup
 */
#ifdef SUNOS4
#define WAITPID		/* use waitpid() instead of wait() */
#endif

#ifdef	SYSLOG
#include <syslog.h>
# ifndef	LOG_AUTH
#  define	LOG_AUTH	0
# endif
# ifndef	LOG_CONS
#  define	LOG_CONS	0
# endif
#endif

#include <ctype.h>
#include <signal.h>
#include <sys/types.h>
#include <stdio.h>
#include <pwd.h>
#include <errno.h>
#ifdef WAITPID
#include <sys/wait.h>
#endif
#include "version.h"

#ifndef lint
static char sccsid[] = "@(#)npasswd.c	1.18 12/20/91 (cc.utexas.edu) /tmp_mnt/usr/share/src/private/ut/share/bin/passwd/SCCS/s.npasswd.c";
#endif

#ifndef	CONFIG_FILE
#define	CONFIG_FILE	"/usr/adm/passwd.conf"
#endif
#ifndef	HELP_FILE
#define	HELP_FILE	"/usr/adm/passwd.help"
#endif
#ifndef	MOTD_FILE
#define	MOTD_FILE	"/usr/adm/passwd.motd"
#endif

extern int	errno;			/* System error code */

#ifdef	sun
static char	*options = "alyd:e:F:n:fPsRVx:";/* Command line options */
#else
static char	*options = "fPRsV";	/* Command line options */
#endif
static int	retries = 3;		/* Retry limit */
static int	from_prog = 0;		/* Data source is a program */
static int	need_oldpw = 0;		/* Require old passwd */

static char	username[16],	/* Name of user changing password */
		password[16];	/* Current password (encrypted) */

char	pbuf[16],		/* Password read buffer 1 */
	pbuf2[16],		/* Password read buffer 2 */
	ppbuf[16],		/* Current password */
	mylogin[16];		/* My login name (saved) */

char	*getpass(),
	*malloc(),
	*ttyname(),
	*getlogin();

#ifdef	SYSV
#define	index	strchr
#endif
char	*index();

int	catchit();		/* Signal catcher */

/*
 *	passwd - change the password for a user.
 *
 *	This program impliments the 'passwd' command.
 */
main(argc, argv)
int	argc;
char	*argv[];
{
	char	*myname, mysavedname[16];/* My login name */
	struct passwd *pw;		/* My passwd entry */
	int	opt;			/* Option processing temp */
	extern char	*optarg;	/* From getopt() */
	extern int	optind;		/* From getopt() */

	/*
	 * Handle the 4.3BSD & SunOS 4.0 command line options.
	 * Defer everything except password change to other programs.
	 */
	while ((opt = getopt(argc, argv, options)) != EOF) {
		switch (opt) {
#ifdef	sun
		/*
		 * Recognized the SunOS 4.1 switches
		 * but just say that we don't handle them.
		 */
		case 'a':
			printf("Option \"-a\" not supported.\n");
			exit(1);
		case 'l':
			printf("Option \"-l\" not supported.\n");
			exit(1);
		case 'y':
			printf("Option \"-y\" not supported.\n");
			exit(1);
		case 'd':
			printf("Option \"-d\" not supported.\n");
			exit(1);
		case 'e':
			printf("Option \"-e\" not supported.\n");
			exit(1);
		case 'n':
			printf("Option \"-n\" not supported.\n");
			exit(1);
		case 'x':
			printf("Option \"-x\" not supported.\n");
			exit(1);
		case 'F':
			printf("Option \"-F\" not supported.\n");
			exit(1);
#endif
		case 'f':
			punt("chfn");
			break;
		case 's':
			punt("chsh");
			break;
		case 'P':		/* Data source is a program */
			if (getuid())
				quit(0, "Option \"-P\" reserved for super-user.\n");
			from_prog = 1;
			break;
		case 'R':		/* Require old passwd even for su */
			if (getuid())
				quit(0, "Option \"-R\" reserved for super-user.\n");
			need_oldpw = 1;
			break;
		case 'V':
			printf("%s; patch level %s\n", version, patchlevel);
			exit(0);
		}
	}
	bzero(ppbuf, sizeof(ppbuf));
#ifndef	DEBUG
	if (geteuid())
		quit(0,"Permission denied.\n");
#endif
	checktty();
	savetty();
	myname = getlogin();
	if (myname == 0 || *myname == '\0') {
		if ((pw = getpwuid(getuid())) == ((struct passwd *)NULL))
			quit(1, "Cannot get your login name.\n");
		strncpy(mysavedname, pw->pw_name, sizeof(mysavedname));
		myname = mysavedname;
	}
	(void) strcpy(mylogin, myname);
	(void) signal(SIGINT, catchit);
	(void) signal(SIGQUIT, catchit);

#ifdef	SYSLOG
	openlog("passwd", LOG_PID | LOG_CONS, LOG_AUTH);
#endif
	setcheckpasswd("-c", CONFIG_FILE, 0);
	pw_initialize(argc, argv);

	if (argv[optind])
		(void) strcpy(username, argv[optind]);
	else
		(void) strcpy(username, mylogin);

	if (strcmp(username, mylogin) == 0 && getuid()) {
		if (pw_getuserbyname(username, password) == 0)
			quit(1, "Cannot get your password information.\n");
		if (password[0])
			getpassword(password, ppbuf, sizeof(ppbuf));
	}
	else {
		if (pw_getuserbyname(username, password) == 0)
			quit(0, "No such user %s\n", argv[optind]);
		if (pw_permission() == 0)
			quit(0, "Permission denied.\n");
		if (need_oldpw)
			getpassword(password, ppbuf, sizeof(ppbuf));
		printf("Changing password for %s\n", username);
	}
	fflush(stdout);
	motd(MOTD_FILE, (char *)0);

	for (;;) {
		char	*px;		/* Temp */
		int	ntries = 0;	/* Password match counter */

		px = getpass(from_prog ? "" : "New password (? for help): ");
		if (px == NULL)
			quit(0, "EOF during new password read.\n");
		(void) strcpy(pbuf, px);
		if (strcmp(pbuf, "?") == 0) {
			motd(HELP_FILE, "Missing help file");
			continue;
		}
		/* Sanity check the new password */
		if (pw_check(pbuf) == 0)
			continue;

		/* Get confirmation */
		px = getpass(from_prog ? "" : "New password (again): ");
		if (px == NULL)
			quit(0, "EOF during new password read.\n");
		(void) strcpy(pbuf2, px);
		if (strcmp(pbuf, pbuf2)) {
			if (ntries++ >= retries) 
				quit(0, "Too many attempts.\n");
			else
				printf("They don't match; try again.\n");
			if (from_prog)
				quit(0, (char *)0);
			else
				continue;
		}
		/* Disallow new password == old password */
		if (pw_compare(password, pbuf)) {
			printf("New password must be different than old; try again.\n");
			if (from_prog)
				quit(0, (char *)0);
			else
				continue;
		}
		else
			break;
	}
	pw_replace(pbuf, ppbuf);
#ifdef	SYSLOG
	syslog(LOG_INFO, "Password changed for %s by %s\n",
		username, mylogin);
#endif
	printf("Password changed for %s\n", username);
	pw_cleanup(0);
	exit(0);
}


/*
 *	getpassword -- read password and check against current.
 */
getpassword(pwd_crypt, pwd_plain, pwlen)
char	*pwd_crypt,		/* Present password (encrypted) */
	*pwd_plain;		/* Present password (plain)  */
int	pwlen;			/* Length of present password buffer */
{
	int	ntries = 0;	/* Match attempt counter */
	char	*px;		/* Temp */

	for (;;) {
		px = getpass(from_prog ? "" : "Current password: ");
		if (px == 0)
			quit(0, "EOF during password read.\n");
		if (*px == '\0')
			continue;
		if (!pw_compare(pwd_crypt, px)) {
			printf("Password incorrect.\n");
			if (ntries++ == retries)
				quit(0, "Password not matched.\n");
		}
		else
			break;
	}
	if (pwd_plain)
		(void) strncpy(pwd_plain, px, pwlen);
}

/* 
 *	randomstring - create a string of random characters
 */
randomstring(buf, len)
char	buf[];		/* String buffer */
int	len;		/* Length of buf */
{
	register int	i,		/* Temp */
			n;		/* Temp */
	time_t		tv;		/* Current time */
	char		proto[128];	/* Build buffer */

	(void) time (&tv);
	/*
	 * Assumes (implicitly) that sizeof(int) == sizeof(long)
	 */
	(void) srand ( (tv & 0x38d9fcff) ^ getpid ());
	for (i = 0; i < sizeof(proto); i++) {	/* fill proto vector */
		int	c;		/* Temp */

		for (;;) {
			c = rand () % 0x7f;	/* turn into ASCII */
			if (isalnum (c))
				break;
		}
		proto[i] = (char )c;
	}
	(void) srand(((unsigned )tv & 0x1a90fefc) ^ getpid());
	for (i = 0; i < len; i++) {
		n = rand() % sizeof(proto);
		buf[i] = proto[n];
	}
	buf[len] = 0;
}

/*
 *	quit - print/log error message and exit
 */
quit(logit, message, a1, a2, a3, a4, a5, a6, a7, a8, a9, a10)
/*VARARGS2*/
int	logit;		/* 0 = don't log, <> 0 = log message */
char	*message;	/* Message */
int	*a1, *a2, *a3, *a4, *a5, *a6, *a7, *a8, *a9, *a10;	/* Args */
{
	if (message) {
		/*
		 * If used from program, direct failure messages to stdout,
		 * else send to stderr.
		 */
		fprintf(from_prog ? stdout : stderr,  message,
			a1, a2, a3, a4, a5, a6, a7, a8, a9, a10);
#ifdef	SYSLOG
		if (logit)
			syslog(LOG_ERR,  message,
				a1, a2, a3, a4, a5, a6, a7, a8, a9, a10);
#endif
	}
	pw_cleanup(1);
	exit(1);
}

/*
 *	motd - issue 'message of the day'
 */
motd(fn, complaint)
char	*fn,			/* Name of file to present */
	*complaint;		/* Complaint if missing */
{
	char	cmdbuf[BUFSIZ];		/* Buffer to build command in */
	int	pid;

	if (access(fn, 0) < 0) {
		if (complaint)
			printf("%s\n", complaint);
		return;
	}
	if (isatty(0))
#ifdef	SYSV
		(void) sprintf(cmdbuf, "pg -n -s %s", fn);
#else
		(void) sprintf(cmdbuf, "more -d %s", fn);
#endif
	else
		(void) sprintf(cmdbuf, "cat %s", fn);

	if ((pid = fork()) == 0) {
		(void) setgid(getgid());
		(void) setuid(getuid());
		(void) system(cmdbuf);
		exit(0);
	}
#ifdef WAITPID
	(void) waitpid(pid, (int *)0, 0); /* wait for child, not child's shell*/
#else
	(void) wait((int *)0);		/* "Wrong" for BSD, right for SYS V */
#endif
}

/*
 *	checktty - Attempt to check against being pipe-fed
 */
checktty()
{
	char	*stdin_tty,	/* ttyname(0) */
		*stdout_tty,	/* ttyname(1) */
		*t;		/* Temp */

	if (!isatty(0)) {
		if (from_prog == 0)
			quit(0, "Input not a tty.\n");
		if (lseek(0, 0L, 1) < 0) {
			if (errno != ESPIPE)
				quit(0, "Input not a tty or pipe.\n");
			else
				return;
		}
	}
	from_prog = 0;		/* Stdin is a tty - behave normal */
	stdin_tty = ttyname(0);
	if (stdin_tty == NULL || *stdin_tty == 0)
		quit(0, "Cannot get name (stdin).\n");
	t = malloc(strlen(stdin_tty) + 1);
	if (t == NULL)
		quit(1, "Cannot allocate temp memory.");
	(void) strcpy(t, stdin_tty);
	stdin_tty = t;

	stdout_tty = ttyname(1);
	if (stdout_tty == NULL || *stdout_tty == 0)
		quit(0, "Cannot get name (stdout).\n");
	if (strcmp(stdin_tty, stdout_tty))
		quit(0, "Input and output are not the same tty.\n");
	free(stdin_tty);
}

/*
 *	catchit - tty interrupt catcher
 */
catchit()
{
	fixtty();
	pw_cleanup(1);
	quit(0, "\nInterrupted; changes discarded.\n");
}


#if	defined(SYSV)
#	include <termio.h>		/* Vanilla SV termio */
	struct termio saved_tty_mode;
#endif

#if	defined(SUNOS4)
#	include <sys/termios.h>		/* SUN OS 4.0 termio */
#define	TCGETA	TCGETS
#define	TCSETA	TCSETS
	struct termios saved_tty_mode;
#endif

#if	!defined(SUNOS4) && !defined(SYSV)
#	include <sgtty.h>		/* BSD tty */
	struct sgttyb saved_tty_mode;
	int	saved_local_flags;
#endif
char	saves_valid  = 0;		/* Are the saved values valid? */

/*
 *	savetty - save current terminal settings
 */
savetty()
{
#if	defined(SYSV) || defined(SUNOS4)
	(void) ioctl(0, TCGETA, &saved_tty_mode);
#else
	(void) ioctl(0, TIOCGETP, &saved_tty_mode);
	(void) ioctl(0, TIOCLGET, &saved_local_flags);
#endif
	saves_valid++;
}

/*
 *	fixtty - restore saved terminal settings
 */
fixtty()
{
	if (saves_valid) {
#if	defined(SYSV) || defined(SUNOS4)
		(void) ioctl(0, TCSETA, &saved_tty_mode);
#else
		(void) ioctl(0, TIOCSETP, &saved_tty_mode);
		(void) ioctl(0, TIOCLSET, &saved_local_flags);
#endif
	}
}

#ifdef	XGETPASS
/*
 *	The system getpass() throws away all but the first 8 characters
 *	of a password string.  If this isn't enough for you, use this
 *	routine instead.  This code assumes that stdin is the terminal.
 */
char	*
getpass(prompt)
char	*prompt;
{
#if	defined(SYSV)
	struct termio	saved,		/* Saved tty characteristics */
			noecho;		/* No-echo tty characteristics */
	char	*strchr();
#endif
#if	defined(SUNOS4)
	struct termios	saved,		/* Saved tty characteristics */
			noecho;		/* No-echo tty characteristics */
#else
	struct sgttyb	saved,		/* Saved tty characteristics */
			noecho;		/* No-echo tty characteristics */
#endif
	static char	ib[64];		/* Input buffer */
	char	*rc;			/* Temp */

#if	defined(SYSV) || defined(SUNOS4)
	(void) ioctl(0, TCGETA, &saved);
	noecho = saved;
	noecho.c_lflag &= ~ECHO;
	(void) ioctl(0, TCSETA, &noecho);
#else
	(void) ioctl(0, TIOCGETP, &saved);
	noecho = saved;
	noecho.sg_flags &= ~ECHO;
	(void) ioctl(0, TIOCSETP, &noecho);
#endif
	fprintf(stderr, "%s", prompt);
	fflush(stderr);
	rc = fgets(ib, sizeof(ib), stdin);
	putc('\n', stderr);
	fflush(stderr);

#if	defined(SYSV) || defined(SUNOS4)
	(void) ioctl(0, TCSETA, &saved);
#else
	(void) ioctl(0, TIOCSETP, &saved);
#endif
	if (rc == NULL)
		return(NULL);
	if (rc = index(ib, '\n'))
		*rc = 0;
	return(ib);
}
#endif

#ifdef	XPUTPWENT
/*
 *	putpwent - replacement for the System V routine
 *		This writes the "standard" passwd file format.
 */
putpwent(p, f)
struct passwd	*p;	/* Passwd entry to put */
FILE	*f;		/* File pointer */
{
#ifdef	UNSIGNED_UID
	fprintf(f, "%s:%s:%u:%u:%s:%s:%s\n",
#else
	fprintf(f, "%s:%s:%d:%d:%s:%s:%s\n",
#endif
		p->pw_name, p->pw_passwd, p->pw_uid, p->pw_gid,
		p->pw_gecos, p->pw_dir, p->pw_shell);
}
#endif

#ifdef	XFGETPWENT
/*
 *	fgetpwent() - read passwd(5) entry from a file
 *		This reads the "standard" passwd file format.
 */
struct passwd *
fgetpwent(f)
FILE	*f;			/* Pointer to open passwd format file */
{
	static struct passwd	pwdata;	/* Return data */
	static char	ibuf[BUFSIZ];	/* Input and return data buffer */
	char		*p;		/* ACME Pointer Works, Inc */

	bzero((char *)&pwdata, sizeof(pwdata));
	pwdata.pw_name = pwdata.pw_passwd = pwdata.pw_comment =
	pwdata.pw_gecos = pwdata.pw_dir = pwdata.pw_shell = "";

	if (fgets(ibuf, sizeof(ibuf), f) == NULL)
		return(0);
	if ((p = index(ibuf, '\n')) == 0)		/* Zap newline */
		quit(1, "Ill-formed passwd entry \"%s\".\n", ibuf);
	else
		*p = 0;
#define	skipc while (*p && *p != ':' && *p != '\n') ++p; if (*p) *p++ = 0
	p = ibuf;
	pwdata.pw_name = p;	skipc;
	pwdata.pw_passwd = p;	skipc;
	pwdata.pw_uid = atoi(p); skipc;
	pwdata.pw_gid = atoi(p); skipc;
	pwdata.pw_gecos = p;	skipc; 
	pwdata.pw_dir = p;	skipc;
	pwdata.pw_shell = p;
	return(&pwdata);
#undef	skipc
}
#endif

#ifdef	SYSV
/*
 *	rename - replacement for the 4.2/4.3 BSD rename system call
 */
rename(src, dst)
char	*src,		/* Source path */
	*dst;		/* Destination path */
{
	if (unlink(dst) < 0) {
		if (errno != ENOENT)
			return(-1);
	}
	if (link(src, dst) < 0)
		return(-1);
	return(unlink(src));
}
#endif

/*
 *	punt() - run another program to do what we don't do
 */
punt(prog)
char	*prog;		/* Program to run */
{
	(void) setgid(getgid());
	(void) setuid(getuid());
	(void) execlp(prog, prog, 0);
	perror(prog);
	exit(1);
}
/*		End npasswd.c		*/
